src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -216,6 +216,53 @@ _mongoc_topology_scanner_cb (uint32_t id,
    bson_mutex_unlock (&topology->mutex);
 }
 
+static void
+_server_session_init (mongoc_server_session_t *session,
+                      mongoc_topology_t *unused,
+                      bson_error_t *error)
+{
+   _mongoc_server_session_init (session, error);
+}
+
+static void
+_server_session_destroy (mongoc_server_session_t *session,
+                         mongoc_topology_t *unused)
+{
+   _mongoc_server_session_destroy (session);
+}
+
+static int
+_server_session_should_prune (mongoc_server_session_t *session,
+                              mongoc_topology_t *topo)
+{
+   bool is_loadbalanced;
+   int timeout;
+
+   /** If "dirty" (i.e. contains a network error), it should be dropped */
+   if (session->dirty) {
+      return true;
+   }
+
+   /** If the session has never been used, it should be dropped */
+   if (session->last_used_usec == SESSION_NEVER_USED) {
+      return true;
+   }
+
+   /* Check for a timeout */
+   bson_mutex_lock (&topo->mutex);
```

Comment:
Just for my own understanding, this was the mutex in question that had all the contention, right?

Reply:
Yes, this is the troublesome mutex. Some of the usages (such as this one being commented on) can be replaced with atomic-loads, which will further alleviate pressure on this lock.